[RDY] Keep unnamed register on restart

AdnoC opened the pull request:
Resolves https://github.com/neovim/neovim/issues/4645

Sets the unnamed register to the last-used one.

Adds a field `is_unnamed` to the ShaDa file for registers and sets the field to whatever register the unnamed register is set to.

Also added an option to `setreg` that sets the unnamed register to the set register (this was needed for testing).

~~Does this by finding the register with the most recent timestamp. If two registers have the same timestamp the first loaded one is used as the unnamed register.~~

~~I did it this way since I didn't really want to mess to much with the ShaDa file.~~

jamessan committed:
shada: Remember whether "0 or "1 was the unnamed register

Ref #4645

jamessan commented:
Looks good to me.  I think @ZyX-I did a lot of the shada work, so his input would be useful.

ZyX-I commented on test/functional/shada/registers_spec.lua:
```diff
@@ -1,6 +1,7 @@
 local helpers = require('test.functional.helpers')
-local nvim_command, funcs, eq = helpers.command, helpers.funcs, helpers.eq
+local nvim_command, funcs, eq, execute = helpers.command, helpers.funcs,
+   helpers.eq, helpers.execute
```
Remove `execute`. There is `nvim_command`, use it.

ZyX-I commented on src/nvim/ops.c:
```diff
@@ -5704,3 +5704,19 @@ const yankreg_T *op_register_get(const char name)
+/// @param[in]  name  Register name.
+///
+/// @return true on success, false on failure.
+bool op_previous_register_set(const char name)
```
That is better named `op_register_set_previous`.

ZyX-I commented:
This actually changes the behaviour: Vim marks some register as the previous one (and so under some circumstances there may be no `"` set: e.g.

    vim -u NONE -i /tmp/viminfo -N --cmd 'call setreg("1", ["abc"])' --cmd 'q!'

). This variant always sets `"` to something, and that something may be different from Vim. Variant that does the same thing that Vim does is something like

```Patch
diff --git a/src/nvim/shada.c b/src/nvim/shada.c
index 32a02b0..6d4f2fd 100644
--- a/src/nvim/shada.c
+++ b/src/nvim/shada.c
@@ -103,6 +103,7 @@ KHASH_SET_INIT_STR(strset)
 #define REG_KEY_TYPE "rt"
 #define REG_KEY_WIDTH "rw"
 #define REG_KEY_CONTENTS "rc"
+#define REG_KEY_UNNAMED "ru"
 
 #define KEY_LNUM "l"
 #define KEY_COL "c"
@@ -287,6 +288,7 @@ typedef struct {
       char **contents;
       size_t contents_size;
       size_t width;
+      bool is_unnamed;
       dict_T *additional_data;
     } reg;
     struct global_var {
@@ -479,6 +481,7 @@ static const ShadaEntry sd_default_values[] = {
           .contents = NULL,
           .contents_size = 0,
           .width = 0,
+          .is_unnamed = false,
           .additional_data = NULL),
   DEF_SDE(Variable, global_var,
           .name = NULL,
@@ -1427,7 +1430,7 @@ static void shada_read(ShaDaReadDef *const sd_reader, const int flags)
           .y_width = (colnr_T) cur_entry.data.reg.width,
           .timestamp = cur_entry.timestamp,
           .additional_data = cur_entry.data.reg.additional_data,
-        })) {
+        }, cur_entry.data.reg.is_unnamed)) {
           shada_free_shada_entry(&cur_entry);
         }
         // Do not free shada entry: its allocated memory was saved above.
@@ -1868,6 +1871,7 @@ static ShaDaWriteResult shada_pack_entry(msgpack_packer *const packer,
           2  // Register contents and name
           + ONE_IF_NOT_DEFAULT(entry, reg.type)
           + ONE_IF_NOT_DEFAULT(entry, reg.width)
+          + ONE_IF_NOT_DEFAULT(entry, reg.is_unnamed)
           // Additional entries, if any:
           + (size_t) (entry.data.reg.additional_data == NULL
                       ? 0
@@ -1888,6 +1892,14 @@ static ShaDaWriteResult shada_pack_entry(msgpack_packer *const packer,
         PACK_STATIC_STR(REG_KEY_WIDTH);
         msgpack_pack_uint64(spacker, (uint64_t) entry.data.reg.width);
       }
+      if (!CHECK_DEFAULT(entry, reg.is_unnamed)) {
+        PACK_STATIC_STR(REG_KEY_UNNAMED);
+        if (entry.data.reg.is_unnamed) {
+          msgpack_pack_true(spacker);
+        } else {
+          msgpack_pack_false(spacker);
+        }
+      }
       DUMP_ADDITIONAL_DATA(entry.data.reg.additional_data, "register item");
       break;
     }
@@ -2741,7 +2753,8 @@ static ShaDaWriteResult shada_write(ShaDaWriteDef *const sd_writer,
     do {
       yankreg_T reg;
       char name = NUL;
-      reg_iter = op_register_iter(reg_iter, &name, &reg);
+      bool is_unnamed = false;
+      reg_iter = op_register_iter(reg_iter, &name, &reg, &is_unnamed);
       if (name == NUL) {
         break;
       }
@@ -2761,6 +2774,7 @@ static ShaDaWriteResult shada_write(ShaDaWriteDef *const sd_writer,
               .width = (size_t) (reg.y_type == MBLOCK ? reg.y_width : 0),
               .additional_data = reg.additional_data,
               .name = name,
+              .is_unnamed = is_unnamed,
             }
           }
         }
@@ -3769,6 +3783,7 @@ shada_read_next_item_start:
             entry->data.reg.contents[i] = BIN_CONVERTED(arr.ptr[i].via.bin);
           }
         }
+        BOOLEAN_KEY("register", REG_KEY_UNNAMED, entry->data.reg.is_unnamed)
         TYPED_KEY("register", REG_KEY_TYPE, "an unsigned integer",
                   entry->data.reg.type, POSITIVE_INTEGER, u64, TOU8)
         TYPED_KEY("register", KEY_NAME_CHAR, "an unsigned integer",
diff --git a/runtime/autoload/shada.vim b/runtime/autoload/shada.vim
index 9be85b6..e885e16 100644
--- a/runtime/autoload/shada.vim
+++ b/runtime/autoload/shada.vim
@@ -45,7 +45,7 @@ call map(copy(s:SHADA_ENTRY_NAMES),
 let s:SHADA_MAP_ENTRIES = {
   \'search_pattern': ['sp', 'sh', 'ss', 'sb', 'sm', 'sc', 'sl', 'se', 'so',
   \                   'su'],
-  \'register': ['n', 'rc', 'rw', 'rt'],
+  \'register': ['n', 'rc', 'rw', 'rt', 'ru'],
   \'global_mark': ['n', 'f', 'l', 'c'],
   \'local_mark': ['f', 'n', 'l', 'c'],
   \'jump': ['f', 'l', 'c'],
@@ -139,6 +139,7 @@ let s:SHADA_STANDARD_KEYS = {
   \'rt': ['type', 'regtype', s:SHADA_ENUMS.regtype.CHARACTERWISE],
   \'rw': ['block width', 'uint', 0],
   \'rc': ['contents', 'binarray', s:SHADA_REQUIRED],
+  \'ru': ['is unnamed', 'boolean', g:msgpack#false],
   \'n':  ['name', 'intchar', char2nr('"')],
   \'l':  ['line number', 'uint', 1],
   \'c':  ['column', 'uint', 0],
```

(I did not provide required changes to tests and ops.c).

AdnoC commented:
Thanks@ZyX-I. I would never have figured out how to add that to the ShaDa file myself.

AdnoC committed:
shada: Set the unnamed register to the previous unnamed register on startup

AdnoC committed:
eval: Add ability to set the unnamed register with setreg

AdnoC committed:
eval/shada: Add testing for unnamed register with setreg and startup

ZyX-I commented:
Also needs documentation update: some file in runtime/doc contains complete ShaDa file description.

ZyX-I commented on test/functional/plugin/shada_spec.lua:
```diff
@@ -715,6 +722,7 @@ describe('In autoload/shada.vim', function()
         '  # Unexpected enum value: expected one of 0 (CHARACTERWISE), '
         .. '1 (LINEWISE), 2 (BLOCKWISE)',
         '  + rt   type         10',
+        '  + ru   is_unnamed   FALSE',
```
Note what tests are testing. Just adding this line to the failing tests is not enough.

ZyX-I replied:
Specifically I remember that I check each and every possible key with unexpected type. Also a test where header contains every single key which is defined for any kind of entry. And a test where header contains every single key which is defined for any kind of entry, but with value of another type.

ZyX-I replied:
Also there are register entry tests which a) contain each key value equal to default one and b) contain each key value *not* equal to default (but yet with a correct type).

AdnoC replied:
What exactly do I need to add for these tests? Is it that I need to test that it errors correctly for each of the non-boolean messagepack types? I only had one with an incorrect type (passed an int instead of a boolean), so should I add more?

I can't find the  case where the values of each key are the default. Can you post the line number?

ZyX-I commented on src/nvim/ops.c:
```diff
@@ -5704,3 +5710,19 @@ const yankreg_T *op_register_get(const char name)
+
+  y_previous = &y_regs[i];
+  return true;
+}
```
This function is no longer needed.

ZyX-I commented on test/functional/shada/registers_spec.lua:
```diff
@@ -160,4 +160,16 @@ describe('ShaDa support code', function()
+    eq({{'zero'}, 'v'}, getreg('0'))
+    eq({{'one'}, 'v'}, getreg('1'))
+    eq({{'one'}, 'v'}, getreg('"'))
+  end)
```
Missing test where unnamed register is not set and thus not restored.

AdnoC replied:
Isn't that what the test directly above it does?

ZyX-I replied:
All tests here test that there something in unnamed register. This does not look like “not set”.

AdnoC replied:
This matches the behavior in Vim as far as I can tell. Doing:

```VimL
call setreg('0', 'This is in reg 0', 'c')
echo getreg('"')
call setreg('3', 'This is in reg 3', 'c')
echo getreg('"')
```

Results in `This is in reg 0` being printed twice.

While I haven't tested the hypothetical case where the viminfo/shada file has something in register `0` but no info on the unnamed register, that won't occur and make a difference unless someone manually edits the file. I have the feeling that the behavior should be the same.

ZyX-I replied:
Just don’t do `setreg('0')`. `setreg('3')` does not set `'"'`.

ZyX-I replied:
The point is that you check what would be done if there are some registers stored in a ShaDa file, but among them there is no one linked with unnamed register. This is achieved by setting `'3'` without setting anything else. Or setting `'a'`. Better set both though.

AdnoC replied:
If the unnamed register isn't set and nothing is stored in register `0`, it will be blank regardless of what the other registers are set to, just like Vim.

AdnoC replied:
Ah. You want a test to test that.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -14755,6 +14757,9 @@ static void f_setreg(typval_T *argvars, typval_T *rettv)
           --stropt;
         }
         break;
+      case 'u': case '"':
```
Different `case`s should be on different lines.

AdnoC commented:
This fails Travis because of a linting error.

The `shada_write` function in [shada.c](https://github.com/AdnoC/neovim/blob/keep-default-register/src/nvim/shada.c#L2422) is over 500 lines long. However, it was originally 499 lines; I only added 3 lines.

ZyX-I commented:
@AdnoC This means you should move something out. I have done this a few times, this is why it is close to 500.

AdnoC committed:
shada/linting: Moved some code out of shada_write.
shada_write was too long (over 500 lines) and caused a linting error.
Register initialization was moved to its own function in order to save lines.

AdnoC commented:
Ummmm... I'm not sure what to do with the log output of the failed Travis job \([this one](https://travis-ci.org/neovim/neovim/jobs/128399189)\). It looks like test49.vim failed due to an error. Because of something?

There aren't any problems when I run `make` inside of src/nvim/testdir (I'm running Mac OS X 10.11.4) and I don't really know how to debug the error in the Travis log.

justinmk commented:
@ZyX-I Anything pending here?

justinmk commented:
@AdnoC Would you be able to rebase (or at least merge-resolve) this?

ZyX-I commented on runtime/doc/starting.txt:
```diff
@@ -1265,29 +1265,32 @@ exactly four MessagePack objects:
+                                                   |NL-used-for-Nul|.
+                       ru   Boolean          false Unnamed register. If this
+                                                   is what the unnamed
+                                                   register is set to.
```
Strange sentence, needs to be fixed.

ZyX-I commented on src/nvim/ops.c:
```diff
@@ -5732,3 +5738,19 @@ const yankreg_T *op_register_get(const char name)
+/// @param[in]  name  Register name.
+///
+/// @return true on success, false on failure.
+bool op_register_set_previous(const char name)
```
FUNC_ATTR_PURE FUNC_ATTR_WARN_UNUSED_RESULT

AdnoC replied:
If we warn on unused result, would it be a good idea to print an error message if it fails? I made it return a bool mainly for the sake of completeness.

AdnoC replied:
Also, it isn't a pure function. It modifies the global variable `y_previous`. When I compiled with `FUNC_ATTR_PURE` it never set `y_previous`. In fact, in gdb it just skipped over calling the function (since I wasn't using the result the compiler probably optimized it away. No reason to call a pure function if you aren't going to use it's return value after all).

ZyX-I replied:
I see. Why this is not void?

AdnoC replied:
Because it can fail (if `name` is an invalid register name) and for completeness' sake I wanted it to communicate that.

ZyX-I commented on src/nvim/ops.c:
```diff
@@ -5704,18 +5705,23 @@ size_t op_register_amount(void)
 /// Set register to a given value
 ///
-/// @param[in]  name  Register name.
-/// @param[in]  reg   Register value.
```
Better save the previous line and make this line have one *less* space: aligning parameter name descriptions makes diffs larger and requires more work (I mean, to reformat if needed), so I no longer do this. AFAIK most other contributors do neither.

ZyX-I commented on src/nvim/shada.c:
```diff
@@ -287,6 +288,7 @@ typedef struct {
       char **contents;
       size_t contents_size;
       size_t width;
+      bool is_unnamed;
```
AFAIK this better be placed after `char`. Here it increases size by `sizeof(void *)` (next pointer must be aligned), but putting it after `char` while both `char` and `bool` do not require to be aligned should not even alter the total size of the `struct reg`.

jamessan replied:
It doesn't look like all of the other structs are optimally packed either.

ZyX-I commented on src/nvim/shada.c:
```diff
@@ -2402,6 +2413,46 @@ static inline ShaDaWriteResult shada_read_when_writing(
   return ret;
 }
 
+/// Initializes registers for writing to the ShaDa file
```
“Initialize”, not “initializes”.

ZyX-I commented on src/nvim/shada.c:
```diff
@@ -2402,6 +2413,46 @@ static inline ShaDaWriteResult shada_read_when_writing(
+/// Initializes registers for writing to the ShaDa file
+///
+/// @param[in]  wms  The WriteMergerState used when writing.
+/// @param[in]  max_reg_lines The maximum number of register lines.
```
Too few spaces before `The`.

ZyX-I commented on src/nvim/shada.c:
```diff
@@ -2402,6 +2413,46 @@ static inline ShaDaWriteResult shada_read_when_writing(
+/// @param[in]  wms  The WriteMergerState used when writing.
+/// @param[in]  max_reg_lines The maximum number of register lines.
+static void shada_initialize_registers(WriteMergerState *const wms,
+                                       int max_reg_lines) {
```
Should be

```C
static inline void shada_initialize_registers(WriteMergerState *const wms
                                                  const int max_reg_lines)
  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_ALWAYS_INLINE
{
```

: missing inlining (optional, but better have it), NONNULL_ALL attribute and `{` must be on its own line.

AdnoC replied:
When I add `FUNC_ATTR_ALWAYS_INLINE` the compiler gives a warning ` always_inline function might not be inlinable [-Wattributes]`

ZyX-I commented on test/functional/plugin/shada_spec.lua:
```diff
@@ -622,6 +622,7 @@ describe('In autoload/shada.vim', function()
         '  # Required key missing: rc',
```
> …
> Also a test where header contains every single key which is defined for any kind of entry.
> And a test where header contains every single key which is defined for any kind of entry, but with value of another type.

Cannot find the third (guess I was wrong: next check looks like it, but has different purpose and less keys), but the second is on line 173.

AdnoC replied:
?

ZyX-I replied:
??

AdnoC commented:
@jamessan if this pull request is still usefull I'll work on this when I get home.

Wish I hadn't forgotten about this PR, I really need to configure my Github alert settings.

jamessan commented:
> if this pull request is still usefull I'll work on this when I get home.

Yes, it's definitely still useful.  I've been meaning to get back to this, but haven't had time.

AdnoC committed:
style/lint

AdnoC committed:
test: Fix and add cases for unnamed register

Also:

Add ru to shada tests with all keys

Add test for unset unnamed and register 0

AdnoC commented:
So I rebased off of master and fixed all the conflicts. I made most of the requested changes. There are 2 changes I commented on asking for clarity on what was being requested. One change (adding `FUNC_ATTR_PURE` to a function that modifies global state) didn't make sense so I didn't add it.

AdnoC committed:
eval.c: Ignore unnamed register error in f_setreg

The error case is already handled and an appropriate error message is
already printed.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -14880,7 +14880,9 @@ static void f_setreg(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   if (set_unnamed) {
-    op_register_set_previous(regname);
+    if (!op_register_set_previous(regname)) {
+      EMSG("Unable to set previous yank register");
```
Missing error number. Need to be something like `E5xxx`, one of the unoccupied. And I would suggest using `emsgf("E5542: Unable to set previous yank register to '%c' (<%02x>)", regname, (unsigned)regname)`: as far as I understand `regname` is the only possible reason of the inabliity so this variant will provide all needed details. (Though `(<%02x>)` is probably not needed, but I do not remember how different non-printable characters will work there.)

ZyX-I replied:
Oops, both missing the translation: `_("E5542: …")`.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -14820,6 +14822,10 @@ static void f_setreg(typval_T *argvars, typval_T *rettv, FunPtr fptr)
           }
           break;
         }
+        case 'u': case '"': { // unnamed register
```
Two spaces before the comment or it would fail linter.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -14872,6 +14878,12 @@ static void f_setreg(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+
+  if (set_unnamed) {
+    if (!op_register_set_previous(regname)) {
+      EMSG("Unable to set previous yank register");
```
And also note that if you have error you need a test where error is checked unless it is too hard (“too hard” mostly applies to FS errors and races). I have seen lots of bugs in error handling code just because it was not tested; some appeared after the refactoring which did not intend to change functionality.

AdnoC replied:
Thinking about this more, `f_setreg` already prints an error in the case where `op_register_set_previous` would fail (when `regname` isn't a valid name it errors with `E354`). So it would make sense to discard `op_register_set_previous`'s result and not print a separate error message.

ZyX-I commented on src/nvim/shada.c:
```diff
@@ -1800,6 +1803,14 @@ static ShaDaWriteResult shada_pack_entry(msgpack_packer *const packer,
         PACK_STATIC_STR(REG_KEY_WIDTH);
         msgpack_pack_uint64(spacker, (uint64_t) entry.data.reg.width);
       }
+      if (!CHECK_DEFAULT(entry, reg.is_unnamed)) {
```
You have not specified the default. Currently it works only because when using `{ .attr = value }` syntax if some `.attr` is missing it is initialized with zeroes. Search for `DEF_SDE(Register` above.

AdnoC replied:
Added a default value.

AdnoC committed:
shada: Add default value for is_unnamed

ZyX-I commented on src/nvim/shada.c:
```diff
@@ -2318,6 +2330,48 @@ static inline void add_search_pattern(PossiblyFreedShadaEntry *const ret_pse,
+        .timestamp = reg.timestamp,
+        .data = {
+          .reg = {
+            .contents = (char **) reg.y_array,
```
This would not pass linter either: need no space after a cast. Note that file was written before linter check in question was introduced and I was personally more fond of the space in most cases. (Also note that `(PossiblyFreedShadaEntry) {` above is not a cast, it needs not be changed.)

AdnoC replied:
That was copy-pasted from the body of another function, but I guess I might as well change it.

AdnoC replied:
It seems like it needs the `(PossibyFreedShadaEntry)` to use designated initialization. Removing gives a syntax error.

ZyX-I replied:
I did not mean you need to remove it. I meant that due to it being not a cast you don’t need to remove a space.

AdnoC replied:
Ah

ZyX-I commented on src/nvim/shada.c:
```diff
@@ -2318,6 +2330,48 @@ static inline void add_search_pattern(PossiblyFreedShadaEntry *const ret_pse,
+///
+/// @param[in]  wms  The WriteMergerState used when writing.
+/// @param[in]  max_reg_lines  The maximum number of register lines.
+static void shada_initialize_registers(WriteMergerState *const wms,
```
ALWAYS_INLINE needs `inline` here as well.

ZyX-I commented:
Other then last comments LGTM.

AdnoC commented:
Oops. Forgot to copy something that changed in the copy-pasted section in `shada_initialize_registers` since I copied it.

AdnoC commented:
Not sure why it would fail job spec tests.

justinmk commented:
QB failure is unrelated. @AdnoC is this ready, or was there another test to be added?

AdnoC commented:
It's good. I want to squash related commits together first though. 

Once I do that I'll mark it as [RDY]. I'll do that once I get back from class in 2-3 hours.

AdnoC commented:
Good to go. I mostly left the branches from last-year untouched, but I tried to merge singe-change lint commits.

Before rebasing I made a branch [which can be found here](https://github.com/AdnoC/neovim/tree/keep-default-register-original). Running `git diff keep-default-register..keep-default-register-original` showed there were no differences, so the rebase should have went well.

justinmk commented:
I'll wait for @ZyX-I to change his review to "approved".

AdnoC commented:
Woo! Glad this is finally finished.
